Basic Cargo extension API support

CXuesong opened the pull request:
Resolves #77. See unit tests in `CargoTests.cs` for usage example. Note that the public API in `WikiClientLibrary.Cargo.Linq` may change dramatically in the future.

* [x] `action=cargoquery` request
* [x] LINQ to Cargo MVP (`.Select`, `.Where`, `.OrderBy`)
    * These model annotation attributes are supported
        * `TableAttribute` (`Schema` is not supported)
        * `ColumnAttribute` (`Type` is not supported)
    * Only these Linq methods are supported: `Select`, `Where`, `OrderBy`, `OrderByDescending`, `ThenBy`, `ThenByDescending`.
        * `Select<T>`: You need to return with `new { ... }` expression. Returning a scalar is not supported.
    * Supported unary operator: `!` (`NOT`)
    * Basic binary operators (e.g. +, -, *, /, AND, OR) are supported.
    * Class functions in BCL (e.g. `String.ToUpper`) are not supported.

BrunoBlanes commented on WikiClientLibrary.Cargo/CargoWikiSiteExtensions.cs:
```diff
@@ -0,0 +1,86 @@
+                    action = "cargoquery",
+                    limit = queryParameters.Limit,
+                    tables = string.Join(",", queryParameters.Tables),
+                    fields = queryParameters.Fields != null ? string.Join(",", queryParameters.Fields) : null,
```
Maybe we could check for null here no? Because `fields` is a mandatory parameter and given that Cargo isn't going to give us any useful warning, we might as well throw our own.

RheingoldRiver replied:
Remove the null check here since we're adding the null check above

RheingoldRiver commented on WikiClientLibrary.Cargo/CargoWikiSiteExtensions.cs:
```diff
@@ -0,0 +1,86 @@
+                    where = queryParameters.Where,
+                    group_by = queryParameters.GroupBy,
+                    having = queryParameters.Having,
+                    join_on = queryParameters.JoinOn,
```
This should probably be an array because if you have 3+ tables you'll have a list of join statements

RheingoldRiver commented on WikiClientLibrary.Cargo/CargoWikiSiteExtensions.cs:
```diff
@@ -0,0 +1,86 @@
+                throw new ArgumentOutOfRangeException(nameof(queryParameters) + "." + nameof(queryParameters.Limit));
+            if (queryParameters.Offset < 0)
+                throw new ArgumentOutOfRangeException(nameof(queryParameters) + "." + nameof(queryParameters.Offset));
+            if (queryParameters.Tables == null)
```
Add a check that if len(tables) >= 1 then Join is not null

RheingoldRiver replied:
Add another check that fields is not null, this is a mandatory parameter in Cargo queries

BrunoBlanes commented on WikiClientLibrary.Cargo/CargoWikiSiteExtensions.cs:
```diff
@@ -0,0 +1,86 @@
+                throw new ArgumentOutOfRangeException(nameof(queryParameters) + "." + nameof(queryParameters.Limit));
+            if (queryParameters.Offset < 0)
+                throw new ArgumentOutOfRangeException(nameof(queryParameters) + "." + nameof(queryParameters.Offset));
+            if (queryParameters.Tables == null)
+                throw new ArgumentException("queryParameters.Tables must be specified", nameof(queryParameters) + "." + nameof(queryParameters.Tables));
+            using (site.BeginActionScope(site, nameof(ExecuteCargoQueryAsync)))
```
```suggestion
            if (queryParameters.Tables == null)
                throw new ArgumentException("queryParameters.Tables must be specified", nameof(queryParameters) + "." + nameof(queryParameters.Tables));
            if (queryParameters.Fields == null)
                throw new ArgumentException("queryParameters.Fields must be specified", nameof(queryParameters) + "." + nameof(queryParameters.Tables));
            using (site.BeginActionScope(site, nameof(ExecuteCargoQueryAsync)))
```
Maybe we could check `Fields` for null here no? Because it is a mandatory parameter and given that Cargo isn't going to give us any useful warning, we might as well throw our own.

RheingoldRiver reviewed (changes requested):
Added a couple notes about the api interface

RheingoldRiver commented on WikiClientLibrary.Cargo/CargoWikiSiteExtensions.cs:
```diff
@@ -0,0 +1,86 @@
+                throw new ArgumentOutOfRangeException(nameof(queryParameters) + "." + nameof(queryParameters.Limit));
+            if (queryParameters.Offset < 0)
+                throw new ArgumentOutOfRangeException(nameof(queryParameters) + "." + nameof(queryParameters.Offset));
+            if (queryParameters.Tables == null)
```
Actually there's one other check you can do, if you want - because of how the response is encoded, none of the field name *aliases* can start with an underscore. So for example if you're querying `_pageName` you would have to alias this as `_pageName=Page`. Additionally, `Champion=_Champion` would be illegal. And Cargo will not provide any kind of helpful api error response if the user fails to alias this way. So you could check for that for every field.

CXuesong replied:
Thanks for your reminder! Generally I'd like to make the pre-condition as permissive as enough, and let MW API Server be the single source of truth. Thus for now I will only do the nullability check.

But specifically for Cargo, we may do some postmortem automatically and help user "guess" what could be the cause if we receive DBError / InternalError. This can be covered later.

CXuesong commented:
I think this PR can conclude here and we can start another one so as to address the limitations. I've published the package on NuGet registry: [CXuesong.MW.WikiClientLibrary.Cargo v0.7.4-int.0](https://www.nuget.org/packages/CXuesong.MW.WikiClientLibrary.Cargo/0.7.4-int.0) and you can try it out a bit.

BrunoBlanes commented:
So I tested it for a bit and couldn't get it to work. Is this the only and proper way to call the API?
```c#
// Call .AsAsyncEnumerable to ensure we use async Linq call.
var records = await q.AsAsyncEnumerable().ToListAsync();
```

I noticed this extension method has a reference to `System.Interactive.Async`. I am currently getting the following error on .NET 5:
```
System.IO.FileNotFoundException: 'Could not load file or assembly 'System.Interactive.Async, Version=3.0.0.0, Culture=neutral, PublicKeyToken=94bc3704cddfc263'. The system cannot find the file specified.'
```

So I went ahead and downloaded the package from NuGet and got the following error message:
```
Reference to type 'IAsyncEnumerable<>' claims it is defined in 'System.Interactive.Async', but it could not be found.
```

BrunoBlanes commented:
Ok, so i got it working with this call:
```c#
var response = query.ToAsyncEnumerable().ToList();
var result = await response;
```

BrunoBlanes commented:
So, after testing it for a while I have a couple of questions:

1. I couldn't get `Where()` to work in my code, am I doing something wrong?

```c#
public async Task GetAsync()
{
    await wikiSite.Initialization;
    var context = new QueryContext(wikiSite);
    var query = context.RosterChanges
        .Where(x => x.Date_Sort > new DateTime(2020, 12, 11))
        .OrderBy(x => x.Date_Sort)
        .Select(x => new
        {
            x.Player,
            x.Date_Sort,
        }).Take(100);
    
    var response = query.ToAsyncEnumerable().ToList();
    var result = await response;
}
```

2. Wouldn't it be wise to enforce a limit when none is specified? I hadn't set a limit at first so this method ran beyond my paciance.

CXuesong commented:
> 1. I couldn't get Where() or OrderBy() to work in my code, am I doing something wrong?

It's a bug. I've fixed it and published `v0.7.4-int2`.

> Also, this is a very memory intensive task, there's a lot of GC going on.

Yeah. I've not done profiling so it's very possible that it has significant memory footprint. I'll let it be for now until '20 😂

> I hadn't set a limit at first so this method ran beyond my paciance.

Actually if you do not limit the size of result set in Entity Framework, you'll possible run into the same result. The point is, the public API let you treat the result set as a `IEnumerable` sequence that has unspecified length. When handling such sequence, it's your responsibility to make sure you are not iterating forever (or almost forever).

Under the hood, I'm currently paginating the result set per 10 records, so after every 10 records, WCL will ask server for the next batch. I'm planning to make the pagination size adjustable in the next PR.

In addition, I've introduced a sample project under `\Samples\LinqToCargo` folder just now. You can check it out. It also connects the logs to console so you can see what is going on when you are enumerating the sequence.

Here is some sample output
<details>

```log
dbug: WikiClientLibrary.Sites.WikiSite[0]
      => Leaguepedia | League of Legends Esports Wiki => WikiSite{Leaguepedia | League of Legends Esports Wiki}::ExecuteCargoQueryAsync()
      Invoke Cargo query. Pseudo-query: SELECT T0.Player = Player,T0.Date_Sort = DateSort FROM RosterChanges = T0
       WHERE (T0.Date_Sort > {dt'2020-12-11T00:00:00.0000000'})
       ORDER BY T0.Date_Sort ASC
       OFFSET 0 FETCH 10 ROWS ONLY
dbug: WikiClientLibrary.Sites.WikiSite[0]
      => Leaguepedia | League of Legends Esports Wiki => WikiSite{Leaguepedia | League of Legends Esports Wiki}::ExecuteCargoQueryAsync()
      Sending request 1FDA629400000003, SuppressAccountAssertion=False
   1: { Player = Paz, DateSort = 2020/12/12 上午 12:00:00 }
   2: { Player = Crash, DateSort = 2020/12/12 上午 12:00:00 }
   3: { Player = Ramune, DateSort = 2020/12/12 上午 12:00:00 }
   4: { Player = Gango, DateSort = 2020/12/12 上午 12:00:00 }
   5: { Player = Helper (Kwon Yeong-jae), DateSort = 2020/12/12 上午 12:00:00 }
   6: { Player = Winged, DateSort = 2020/12/12 上午 12:00:00 }
   7: { Player = ScrappyDoo, DateSort = 2020/12/12 上午 12:00:00 }
   8: { Player = Navy (Micha? Leszczyński), DateSort = 2020/12/12 上午 12:00:00 }
   9: { Player = iBo, DateSort = 2020/12/12 上午 12:00:00 }
  10: { Player = Mocha (Kim Tae-gyeom), DateSort = 2020/12/12 上午 12:00:00 }
dbug: WikiClientLibrary.Sites.WikiSite[0]
      => Leaguepedia | League of Legends Esports Wiki => WikiSite{Leaguepedia | League of Legends Esports Wiki}::ExecuteCargoQueryAsync()
      Invoke Cargo query. Pseudo-query: SELECT T0.Player = Player,T0.Date_Sort = DateSort FROM RosterChanges = T0
       WHERE (T0.Date_Sort > {dt'2020-12-11T00:00:00.0000000'})
       ORDER BY T0.Date_Sort ASC
       OFFSET 10 FETCH 10 ROWS ONLY
dbug: WikiClientLibrary.Sites.WikiSite[0]
      => Leaguepedia | League of Legends Esports Wiki => WikiSite{Leaguepedia | League of Legends Esports Wiki}::ExecuteCargoQueryAsync()
      Sending request 1FDA629400000004, SuppressAccountAssertion=False
  11: { Player = Paz, DateSort = 2020/12/12 上午 12:00:00 }
  12: { Player = Crash, DateSort = 2020/12/12 上午 12:00:00 }
  13: { Player = Ramune, DateSort = 2020/12/12 上午 12:00:00 }
  14: { Player = Gango, DateSort = 2020/12/12 上午 12:00:00 }
  15: { Player = Helper (Kwon Yeong-jae), DateSort = 2020/12/12 上午 12:00:00 }
  16: { Player = Winged, DateSort = 2020/12/12 上午 12:00:00 }
  17: { Player = ScrappyDoo, DateSort = 2020/12/12 上午 12:00:00 }
  18: { Player = Navy (Micha? Leszczyński), DateSort = 2020/12/12 上午 12:00:00 }
  19: { Player = iBo, DateSort = 2020/12/12 上午 12:00:00 }
  20: { Player = Mocha (Kim Tae-gyeom), DateSort = 2020/12/12 上午 12:00:00 }
dbug: WikiClientLibrary.Sites.WikiSite[0]
      => Leaguepedia | League of Legends Esports Wiki => WikiSite{Leaguepedia | League of Legends Esports Wiki}::ExecuteCargoQueryAsync()
      Invoke Cargo query. Pseudo-query: SELECT T0.Player = Player,T0.Date_Sort = DateSort FROM RosterChanges = T0
       WHERE (T0.Date_Sort > {dt'2020-12-11T00:00:00.0000000'})
       ORDER BY T0.Date_Sort ASC
       OFFSET 20 FETCH 10 ROWS ONLY
dbug: WikiClientLibrary.Sites.WikiSite[0]
      => Leaguepedia | League of Legends Esports Wiki => WikiSite{Leaguepedia | League of Legends Esports Wiki}::ExecuteCargoQueryAsync()
      Sending request 1FDA629400000005, SuppressAccountAssertion=False
  21: { Player = Paz, DateSort = 2020/12/12 上午 12:00:00 }
  22: { Player = Crash, DateSort = 2020/12/12 上午 12:00:00 }
  23: { Player = Ramune, DateSort = 2020/12/12 上午 12:00:00 }
  24: { Player = Gango, DateSort = 2020/12/12 上午 12:00:00 }
  25: { Player = Helper (Kwon Yeong-jae), DateSort = 2020/12/12 上午 12:00:00 }
  26: { Player = Winged, DateSort = 2020/12/12 上午 12:00:00 }
  27: { Player = ScrappyDoo, DateSort = 2020/12/12 上午 12:00:00 }
  28: { Player = Navy (Micha? Leszczyński), DateSort = 2020/12/12 上午 12:00:00 }
  29: { Player = iBo, DateSort = 2020/12/12 上午 12:00:00 }
  30: { Player = Mocha (Kim Tae-gyeom), DateSort = 2020/12/12 上午 12:00:00 }
…
```

</details>

BrunoBlanes commented:
Yep, working great so far. I'll make sure to test all the possible combinations of queries soon.

One last question for now though, is it possible for an overload of `await site.ExecuteCargoQueryAsync()` to take in an `IQueriable<T>` as parameter and return an `IAsyncEnumerable<T>`?

It is more of a code cleanup for the end user and would resemble more of EF Core's usage.

RheingoldRiver commented:
Hey, for the LINQ support, Cargo has one special operator you can use in `where`, called `HOLDS` - documentation [here](https://www.mediawiki.org/wiki/Extension:Cargo/Querying_data#The_%22HOLDS%22_command), it's just a syntax sugar (and pretty buggy tbh, I try to avoid using it). I didn't look too closely at the PR so you might have added it already, but wanted to link this just in case!

CXuesong commented:
> It is more of a code cleanup for the end user and would resemble more of EF Core's usage.
@BrunoBlanes Actually ExecuteCargoQueryAsync is the counterpart of `SqlCommand` in ADO.NET. It executes a query (in string) and returns the result set. You need to start a IQueryable chained calls from something like an EF `DbSet<T>`, i.e. `ICargoRecordSet` in WCL.

And frankly speaking, the async extension method support for `IQueryable` is not so straightforward in .NET.
1. There is no `IAsyncQueryable`. `IQueryable` implements `IEnumerable` but (perhaps due to compatibility) not `IAsyncEnumerable`, so **you need to restrain yourself from using any of the sync extension method exposed from `IEnumerable`** like this
```c#
context.Skins.First()    // Enumerable.First(IEnumerable<T>)
context.Skins.Take(10).ToArray()    // Enumerable.ToArray(IEnumerable<T>)
```
because we don't want synchronous blocking call in async functions.
2. Because there is no BCL extension method support for `IAsyncEnumerable` (There is no `IAsyncEnumerable<T>.ToArrayAsync` in .NET 5!), WikiClientLibrary is referencing [`System.Linq.Async` (aka. `Ix.Async`)](https://github.com/dotnet/reactive) NuGet package. It's not part of BCL, and if you are consuming this package (via WikiClinetLibrary) and EF at the same time, you need to be careful not to fall into dotnet/efcore#18124 .

Though it may sound a bit frustrating, but I'm not implementing async extension methods like `IQueryable.ToArrayAsync` or `IAsyncEnumerable.ToArrayAsync`, because
1. These methods are implemented in [`efcore`](https://github.com/dotnet/efcore/blob/fc29d1cc809332c5935d13478b988e11e020606e/src/EFCore/Extensions/EntityFrameworkQueryableExtensions.cs#L1989). For now I don't want to repeat these extension methods in WCL.
2. Implementing these extension methods in WCL may dig more name conflict pitfalls like https://github.com/dotnet/efcore/issues/18124#issuecomment-734893944.

This means you will have to use something like this if you want to call some Linq collection (i.e. "collecting") methods
```c#
cargoContext.Skins
    .Take(10)    // IQueryable, no access to ToListAsync
    .AsAsyncEnumerable()    // This function is provided by WikiClientLibrary. It casts the IQueryable to IAsyncEnumerable
                            // Do not confuse it with `ToAsyncEnumerable`!
    .ToListAsync()    // Calls the collection method from Ix.Async
```

If you are also referencing EF Core library, then I suppose the following usage should also do, but I have not ever tried it before.
```c#
cargoContext.Skins
    .Take(10)    // IQueryable, no access to ToListAsync
    .ToListAsync()    // Calls the collection method from EF Core
```

CXuesong commented:
> Hey, for the LINQ support, Cargo has one special operator you can use in where, called HOLDS

@RheingoldRiver Yeah I have that in mind. I think I can implement some stub method like `EF.Functions.Like` in the next PR to make query builder build such query expression.
